platform/engineering/design-docs/2021-08-01-autonomous-deployments.md

```diff
@@ -0,0 +1,197 @@
+
+Upon initial/beta release of production deployments there will be an "allowed" list for applications that VSP identifies as good candidates for testing the system (teams with frequent stable releases). All other applications will have the advantage of auto deployments to staging and dev environments, but will remain contained in the daily deployment until the system is ready to be truly switched on. Applications using the auto-deployment will also remain in the daily deployment while the system is in beta to clean up and issues and dependencies upgrades. This will also reduce complexity.
+
+#### Staging and Dev deployments
+
+Deployments to staging and dev will be automatic upon merge to the branch for all applications. This is a workflow change and will need heavy communication throughout VFS teams. This will mimic the current staging/dev deployment we have in master currently, but no tests will be run. This will cut down on failures due to flakiness in an environment where it isn't really needed since commits still need to complete the build and run through tests in a PR before merging into master in addition to tests run in master CI.
+
+## Specifics
+
+### Detailed Design
+
+**Steps for a production deployment:**
+
+VSP maintains the ability to create and deploy full production builds as a failsafe until the system is stable.
+
+1. Merge feature branch into master to trigger workflow
+2. Is the daily deployment running and/or is master running on this commit?
+  - **Yes**: Pause until finished
+  - **No**: Continue
+3. Has an approved application changed?
+  - **Yes**: Continue
+  - **No**: Stop
+4. Build changed applications / Run unit tests
+5. Sync CSS/JS/images to prod S3 bucket
+
+Manual deployments follow the same process, but do not need to check the commit status. Should be able to trigger via the Github action UI and provide an app name (or list of apps).
+
+**Steps for a staging/dev deployment:**
+
+VSP maintains the ability to create and deploy full staging/dev builds as a failsafe until the system is stable.
+
+1. Merge branch into staging/dev (locally or through PR)
+2. Build all applications
+3. If a deployment to staging/dev in in progress wait in a queue
+4. Sync /build/vagov{env} to staging/dev S3 bucket
+
+This will likely cause `staging` and `dev` to diverge from `master` over time. We can set up a scheduled job to merge master into staging/dev on a weekly basis and reserve the ability to completely reset the environments if needed. When an environment is reset, a slack notification should go out alerting VFS teams that they will need to remerge any branches. An alert should also go out to Slack the previous day if any preparations are needed as to not interfere with potential planned demos. 
+
+#### Roadmap
```

Comment:
Should an education item be included in this roadmap? This will be a big change for our FE teams, I think we should have a significant outreach effort attached to this initiative.

Reply:
That's outline in the milestone section "Plan and execute VFS training on the update"

Reply:
Ah, I see, it's not its own milestone, it's under "Create and release MVP production behaviors with 'disallow' list". Thanks.